File: solutions/2026-06-14/javascript/2026-06-14-maximum-twin-sum-of-a-linked-list.js
/**
 * @param {ListNode} head
 * @return {number}
 */
// Problem: Maximum Twin Sum of a Linked List
// Link: https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list/
// Approach:
// 1. Find the middle of the linked list using the fast and slow pointer technique.
// 2. Reverse the second half of the linked list.
// 3. Iterate through the first half and the reversed second half simultaneously, calculating the twin sum for each pair of nodes.
// 4. Keep track of the maximum twin sum encountered.
// Time Complexity: O(n), where n is the number of nodes in the linked list.
//   - Finding the middle: O(n/2) = O(n)
//   - Reversing the second half: O(n/2) = O(n)
//   - Calculating twin sums: O(n/2) = O(n)
// Space Complexity: O(1), as we are modifying the linked list in-place and using only a few extra pointers.
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
var pairSum = function(head) {
    // Initialize slow and fast pointers to traverse the list.
    let slow = head;
    let fast = head;

    // Move slow pointer one step at a time and fast pointer two steps at a time.
    // When fast reaches the end, slow will be at the middle of the list.
    while (fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
    }

    // Now, 'slow' is pointing to the start of the second half of the linked list.
    // We need to reverse the second half to easily pair nodes with their twins.

    let prev = null;
    let current = slow;
    let nextNode = null;

    // Reverse the second half of the linked list.
    while (current) {
        nextNode = current.next; // Store the next node
        current.next = prev;    // Reverse the current node's pointer
        prev = current;         // Move 'prev' one step forward
        current = nextNode;     // Move 'current' one step forward
    }
    // After the loop, 'prev' will be the head of the reversed second half.

    // Now, 'head' is the start of the first half and 'prev' is the start of the reversed second half.
    // We can now iterate through both halves simultaneously to calculate twin sums.
    let maxTwinSum = 0;
    let firstHalf = head;
    let secondHalfReversed = prev;

    // Iterate as long as both pointers are valid.
    while (secondHalfReversed) {
        // Calculate the sum of the current node from the first half and its twin from the reversed second half.
        const currentTwinSum = firstHalf.val + secondHalfReversed.val;
        // Update the maximum twin sum found so far.
        maxTwinSum = Math.max(maxTwinSum, currentTwinSum);

        // Move both pointers to their next respective nodes.
        firstHalf = firstHalf.next;
        secondHalfReversed = secondHalfReversed.next;
    }

    // Return the maximum twin sum found.
    return maxTwinSum;
};
```

File: solutions/2026-06-14/cpp/2026-06-14-maximum-twin-sum-of-a-linked-list.cpp
```cpp
// Problem: Maximum Twin Sum of a Linked List
// Link: https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list/
//
// Approach:
// The problem asks for the maximum sum of a node and its twin. A twin of the i-th node (0-indexed)
// is the (n-1-i)-th node, where n is the even length of the linked list. This means we need to pair
// the first half of the list with the reversed second half.
//
// We can solve this by:
// 1. Finding the middle of the linked list. We can use the fast and slow pointer technique. The slow
//    pointer will end up at the head of the second half of the list when the fast pointer reaches the end.
// 2. Reversing the second half of the linked list. This will allow us to easily access the twin nodes
//    in a parallel traversal.
// 3. Traversing the first half of the original list and the reversed second half simultaneously.
//    For each pair of nodes (one from the first half, one from the reversed second half), calculate
//    their sum and keep track of the maximum sum found so far.
//
// Time Complexity:
// O(n), where n is the number of nodes in the linked list.
// - Finding the middle: O(n/2) which is O(n).
// - Reversing the second half: O(n/2) which is O(n).
// - Traversing and summing: O(n/2) which is O(n).
//
// Space Complexity:
// O(1). We are only using a few extra pointers for traversal and manipulation, no additional data structures
// that scale with the input size are used.
//

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        // 1. Find the middle of the linked list using fast and slow pointers.
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }
        // 'slow' is now at the head of the second half of the list.

        // 2. Reverse the second half of the linked list.
        ListNode* prev = nullptr;
        ListNode* current = slow;
        ListNode* next_node = nullptr;
        while (current != nullptr) {
            next_node = current->next; // Store the next node
            current->next = prev;      // Reverse the current node's pointer
            prev = current;            // Move pointers one position ahead
            current = next_node;
        }
        // 'prev' is now the head of the reversed second half.

        // 3. Traverse the first half and the reversed second half simultaneously, calculating twin sums.
        ListNode* first_half_ptr = head;
        ListNode* second_half_ptr = prev; // Head of the reversed second half
        int max_twin_sum = 0;

        while (second_half_ptr != nullptr) { // Iterate until the end of the reversed second half
            int current_twin_sum = first_half_ptr->val + second_half_ptr->val;
            max_twin_sum = std::max(max_twin_sum, current_twin_sum);

            first_half_ptr = first_half_ptr->next;
            second_half_ptr = second_half_ptr->next;
        }

        // Note: We don't need to restore the original list structure as per the problem's requirement
        // of just returning the maximum twin sum.

        return max_twin_sum;
    }
};
```